mkPythonMetaPackage: init meta package function (and psycopg2-binary)

adisbladis opened the pull request:
## Description of changes

Adds a function `mkPythonMetaPackage` that exists to satisfy a dependency on a package, but don't actually build it.
This is useful for packages that have a split binary/source dichotomy like psycopg2/psycopg2-binary, where we want to use the former, but some projects declare a dependency on the latter.

This PR:
- Create a `mkPythonMetaPackage` function
- Adds a new package `psycopg2-binary` that uses it

Override hacks could be dropped from:
- irrd
- pretix
- vwsfriend
- aiopg
- dbt-postgres
- llama-index-vector-stores-postgres
- patator
- fit-trackee

cc @misuzu https://github.com/nix-community/pyproject.nix/issues/143

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

adisbladis committed:
python3Packages.psycopg2-binary: init at 2.9.9

Pscycopg2-binary is normally used in Python development to avoid having to build psycopg2 from source.
In nixpkgs we always want ot build from source whenever possible, but it can still be useful to provide a psycopg2-binary package.

This "fake" package exists to satisfy a dependency on psycopg2-binary, but still use the build from psycopg2.

cc @misuzu https://github.com/nix-community/pyproject.nix/issues/143

jian-lin commented on pkgs/applications/editors/emacs/elisp-packages/manual-packages.nix:
```diff
@@ -15,10 +15,6 @@ lib.packagesFromDirectoryRecursive {
     inherit (pkgs) codeium;
   };
 
-  consult-gh = callPackage ./manual-packages/consult-gh {
```
It seems you just pushed the wrong thing.

misuzu reviewed (approved):
Thanks! This seems to work

dotlambda reviewed (commented):
Very nice!
I wonder if it would make sense to turn this into a function to produce such a package that could also be used for e.g. https://pypi.org/project/phonenumberslite/

adisbladis committed:
python3Packages.mkPythonMetaPackage: init

This function exists create a meta package containing [metadata files](https://packaging.python.org/en/latest/specifications/recording-installed-packages/) to satisfy a dependency on a package, without it actually having been installed into the environment.

adisbladis commented:
> Very nice! I wonder if it would make sense to turn this into a function to produce such a package that could also be used for e.g. https://pypi.org/project/phonenumberslite/

Good idea! I've added a `mkPythonMetaPackage` function to the Python set and changed `psycopg2-binary` to use it.

dotlambda commented on pkgs/development/interpreters/python/meta-package.nix:
```diff
@@ -0,0 +1,58 @@
+  dependencies ? [ ],
+  optional-dependencies ? { },
+  passthru ? { },
+  meta ? { },
```
Inheriting `meta` doesn't make sense for the license, does it?

adisbladis replied:
That's a bit of a philosophical question :sweat_smile: 

Either we consider it as `psycopg2-binary` functionality provided by `psycopg2`, under the license terms of `psycopg2`.
Or we leave it empty, just inheriting homepage & description. I'm OK with either.
